pkg/quota/image/usage.go

```diff
@@ -10,139 +13,400 @@ import (
+				if err != nil {
+					glog.Errorf("Failed to get image %s of image stream %s/%s: %v", imageName, is.Namespace, is.Name, err)
+					continue
+				}
+			} else {
+				img, err = c.getImage(imageName)
+				if err != nil {
+					glog.Errorf("Failed to get image %s: %v", imageName, err)
+					continue
+				}
 			}
-			processedImages[imageName] = sets.Empty{}
 
-			imi, err2 := osClient.ImageStreamImages(is.Namespace).Get(is.Name, imageName)
-			if err2 != nil {
-				glog.Errorf("Failed to get image %s of image stream %s/%s: %v", image.Name, is.Namespace, is.Name, err2)
+			if processedImages.Has(imageName) {
+				glog.V(4).Infof("Skipping image %q - already processed", imageName)
 				continue
 			}
-			img := &imi.Image
+			processedImages.Insert(imageName)
 
 			if value, ok := img.Annotations[imageapi.ManagedByOpenShiftAnnotation]; !ok || value != "true" {
 				glog.V(5).Infof("Image %q with DockerImageReference %q belongs to an external registry - skipping", img.Name, img.DockerImageReference)
 				continue
 			}
 
-			if len(img.DockerImageLayers) == 0 || img.DockerImageMetadata.Size == 0 {
-				if err2 := imageapi.ImageWithMetadata(img); err2 != nil {
-					glog.Errorf("Failed to parse metadata of image %q with DockerImageReference %q: %v", img.Name, img.DockerImageReference, err2)
-					continue
-				}
+			cacheInternalRegistryName(img.DockerImageReference)
+
+			if err := handler(tag, history.Items[i].DockerImageReference, img); err != nil {
+				return err
 			}
+		}
+	}
+
+	if c.processSpec {
+		return c.processImageStreamSpecImages(is, processedImages, handler)
+	}
+
+	return nil
+}
+
+// processImageStreamSpecImages is a utility method that calls a given handler on every image of the given image
+// stream that belongs to internal registry. It process image stream's spec only.
+func (c *GenericImageStreamUsageComputer) processImageStreamSpecImages(
+	is *imageapi.ImageStream,
+	processedImages sets.String,
+	handler func(tag string, dockerImageReference string, image *imageapi.Image) error,
+) error {
+	for tag, tagRef := range is.Spec.Tags {
+		if tagRef.From == nil {
+			continue
+		}
+
+		ref, err := c.getImageReferenceForObjectReference(is.Namespace, tagRef.From)
+		if err != nil {
+			glog.V(4).Infof("Could not process object reference: %v", err)
+			continue
+		}
+
+		img, exists := c.imageCache[ref.ID]
+		if !exists {
+			if c.fetchImagesFromImageStream || len(ref.ID) == 0 {
+				if len(ref.ID) > 0 {
+					img, err = c.getImageStreamImage(ref.Namespace, is.Name+"@"+ref.ID)
+					if err != nil {
+						glog.Errorf("Failed to get image stream image %s/%s@%s: %v", ref.Namespace, ref.Name, ref.ID, err)
+						continue
+					}
 
-			for _, layer := range img.DockerImageLayers {
-				if _, exists := processedLayers[layer.Name]; exists {
+				} else {
+					tag = ref.Tag
+					if len(tag) == 0 {
+						tag = "latest"
+					}
+					ist, err2 := c.osClient.ImageStreamTags(ref.Namespace).Get(ref.Name, tag)
```

Comment:
This is incredibly inefficient, and we should never be doing this in this code path.